File: front-end/src/app/model/take.ts
import { QuizDto } from "./quiz-dto";
import { UserQuestionAnswer } from "./user-question-answer";

export class Take {
  id: string;
  userName: string;
  quiz: QuizDto;

  startTime: string;
  endTime: string;
  userQuestionAnswers: UserQuestionAnswer[];
  score: number;
  numberCorrectAnswers: number;
  numberIncorrectAnswers: number;
  timeTaken: number;

  static JsonToTake(json: string) {
    const data = JSON.parse(json);
    console.log(data);
    const quiz: Take = Object.assign(new Take(), data);
    return quiz;
  }

  static convertTimeTakenToString(timeTaken: number) {
    let allSeconds = Math.floor(timeTaken / 1000);
    let hour = Math.floor(allSeconds / 3600);
    let min = Math.floor(allSeconds / 60);
    let sec = Math.round(((allSeconds / 60) - min) * 60);
    return `${hour}:${min}:${sec}`;
  }
}

File: back-end/src/main/java/com/lumberjack/quizappbackend/model/quiz/Take.java
package com.lumberjack.quizappbackend.model.quiz;

import com.lumberjack.quizappbackend.model.quiz.dto.QuizDto;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.io.Serial;
import java.io.Serializable;
import java.util.List;

@Getter
@Setter
@ToString
@NoArgsConstructor
@Document(collection = "quizTakes")
public class Take implements Serializable {
    @Serial
    private static final long serialVersionUID = -4354354354543543552L;

    @Id
    private String id;
    private String userName;
    private QuizDto quiz;
    private String startTime;
    private String endTime;
    private List<UserQuestionAnswer> userQuestionAnswers;
    private double score;
    private int numberCorrectAnswers;
    private int numberIncorrectAnswers;
    private int timeTaken;

    public void setQuiz(Quiz quiz) {
        QuizDto quizDto = new QuizDto();
        quizDto.setId(quiz.getId());
        quizDto.setCategory(quiz.getCategory());
        quizDto.setTitle(quiz.getTitle());
        quizDto.setCreatedAt(quiz.getCreatedAt());
        quizDto.setTimeLimit(quiz.getTimeLimit());
        this.quiz = quizDto;
    }
}
